jaeger_client/throttler.py

```diff
@@ -0,0 +1,199 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import, division
+import json
+import logging
+import random
+from threading import Lock
+
+from tornado.ioloop import PeriodicCallback
+
+from .constants import DEFAULT_THROTTLER_REFRESH_INTERVAL
+from .metrics import Metrics, MetricsFactory
+from .utils import ErrorReporter
+
+MINIMUM_CREDITS = 1.0
+default_logger = logging.getLogger('jaeger_tracing')
+
+
+class RemoteThrottler(object):
+    """
+    RemoteThrottler controls the flow of spans emitted from client to prevent
+    flooding. RemoteThrottler requests credits from the throttling service
+    periodically. These credits determine the amount of debug spans a client
+    may emit for a particular operation without receiving more credits.
+    :param channel: channel for communicating with jaeger-agent
+    :param service_name: name of this application
+    :param kwargs: optional parameters
+        - refresh_interval: interval in seconds for requesting more credits
+        - logger: Logger instance
+        - metrics_factory: factory to create throttler-specific metrics
+        - error_reporter: ErrorReporter instance
+    """
+
+    def __init__(self, channel, service_name, **kwargs):
+        self.channel = channel
+        self.service_name = service_name
+        self.client_id = None
+        self.refresh_interval = \
+            kwargs.get('refresh_interval', DEFAULT_THROTTLER_REFRESH_INTERVAL)
+        self.logger = kwargs.get('logger', default_logger)
+        metrics_factory = kwargs.get('metrics_factory', MetricsFactory())
+        self.metrics = ThrottlerMetrics(metrics_factory)
+        self.error_reporter = kwargs.get('error_reporter', ErrorReporter(Metrics()))
+        self.credits = {}
+        self.lock = Lock()
+        self.running = True
+        self.periodic = None
+
+        if not self.channel.io_loop:
+            self.logger.error(
+                'Cannot acquire IOLoop, throttler will not be updated')
+        else:
+            self.channel.io_loop.add_callback(self._init_polling)
+
+    def is_allowed(self, operation):
+        with self.lock:
+            if operation not in self.credits:
+                self.credits[operation] = 0.0
+                self.metrics.throttled_debug_spans(1)
+                return False
+            value = self.credits[operation]
+            if value < MINIMUM_CREDITS:
+                self.metrics.throttled_debug_spans(1)
+                return False
+
+            self.credits[operation] = value - MINIMUM_CREDITS
+            return True
+
+    def _set_client_id(self, client_id):
+        """
+        Method for tracer to set client ID of throttler.
+        """
+        with self.lock:
+            if self.client_id is None:
```

Comment:
Are you asking or suggesting? I did lock here because to make the interface completely thread-safe. Seeing as the caller is the tracer constructor, I don't really foresee a race condition. What do you think is more correct?